### PR TITLE
Feature/2005  User Management - Functionality Improvements

### LINKED ISSUE #2005: User Management - Functionality Improvements
Improvements to functionality of User Management on the Production Platform:

- [ ] A search box should be added
- [ ] All users should be displayed in alphabetical order
- [ ] Users listed should be only the ones accessing platform via Microsoft log in / using judicialappointments.gov.uk domain
- [ ] All .digital accounts should not be shown in the user list
- [ ] When the list is populated on screen only those mention in point 3 should be searched for in the database to speed up the return on the data

Image

### PR DESCRIPTION
**Pull request by HalcyonJAC**
## What's included?
Closes #2005 

- [x] Create a `users` collection in Firestore for pagination and search.
- [x] Write a node script to populate data from the authentication database into `users` collection.
- [x] Update UI.
- [x] Update sign-in process.
- [x] Add a background function `onUserDelete` to delete the user in the authentication database.
- [x] Add a background function `onUserUpdate` to update the user information (e.g. displayName, email, disabled) in the authentication database.
- [x] Add a background function `onRoleUpdate` to mark users with this role as changed.
- [x] Update the contact dropdown to get the user list from the `users` collection.
- [x] Remove hard-coded email addresses.
- [x] Create a `userInvitations` collection to invite new users.

Note: This PR relies on the changes in [digital-platform: Feature/2005 user management improvements](https://github.com/jac-uk/digital-platform/pull/910).

#### Sign in
1. `auth.onAuthStateChanged(user)` is triggered.
2. Dispatch `auth/setCurrentUser` action.
3. Check if the email is valid.
4. Check if the user document exists in Firestore.
    - No (new user)
        1. Check if there is a user invitation document with the status `pending` in Firestore.
        2. Create a user document in Firestore.
        3. Update the user role information in the custom claims.
        4. Mark the status of the user invitation as `completed`.
5. Ensure the Microsoft email is verified. 
6. Refresh the token using `auth.currentUser.getIdToken(true)` to get the latest custom claims.
7. Get custom claims using `user.getIdTokenResult()`
8. Set the store (`bindCurrentUser`, `setRolePermissions`)

#### A user’s role is changed
Mark `role.isChanged` as `true` in the user document.

[digital-platform] `onUserUpdate`
1. Update role permissions in the custom claims.
5. Mark `role.isChanged` as `false` in the user document.

Note: The currently logged-in users can get new permissions after refreshing the page.

#### Role permissions are changed
Update role permissions in the role document.

[digital-platform] `onRoleUpdate`
1. Get users with this role and mark `role.isChanged` as true.
2. Trigger `onUserUpdate` and update role permissions in the custom claims.

Note: The currently logged-in users can get new permissions after refreshing the page.

#### `role.isChanged` is changed to `true`
Show pop-up to refresh the page.

#### New admin users
1. Create a user invitation for new users.
2. New users log in using their Google/Microsoft account.
3. Check if a user invitation has been created. If a user invitation exists in the Firestore, the user can log in to the platform with the given user role.

## Who should test?
✅ Product owner
✅ Developers

## How to test?

Preview URL: https://jac-admin-develop--pr2054-feature-2005-user-ma-xvbo9tuh.web.app

1. Check if you can sign in to the admin site without problems.

2. Go to `Users` page:
    - Check if all admin users are listed in the table.
    - Check if you can search for a specific user.
    - Check if the pagination of the user table works.
    - Check if you can enable/disable/delete a user.
    - Check if you can change the role.

3. Go to `Roles` page:
    - Check if you can see all roles.
    - Check if you can change the permissions of a role.
    - Check if you can create a new role.

4. Go to `User Invitations` page:
    - Check if you can create a user invitation with an JAC email address and user role.
    - Check if you can use the email to sign in. 

## Risk - how likely is this to impact other areas?
🟠 Medium risk - this does change code that is shared with other areas

## Related permissions
Permissions have been added:
- Can read users `u7`

---
PREVIEW:DEVELOP
_can be OFF, DEVELOP or STAGING_

### CODE REVIEW

File: src/main.js [hunk 1]
```diff
@@ -30,36 +30,25 @@ auth.onAuthStateChanged(async (user) => {
...
-            rolePermissions: permissions,
-          };
-          store.dispatch('auth/setUserRole', userRole);
-        }
+  const userDoc = await store.dispatch('auth/setCurrentUser', user);
+  if (userDoc) {
+    const role = await store.dispatch('roles/get', userDoc.role.id);
+    await store.dispatch('auth/setUserRole', role);
+    if (role) {
+      isAuth = true;
```
Comment (warrensearle) [inline src/main.js:40]: Let's discuss. We need to always be using the auth token / custom claims rather than what is in the `users` document

**Review by warrensearle — CHANGES_REQUESTED** [review]
I have a few questions/changes. Let's go through this via DM 😄

File: src/store/users.js [hunk 2]
```diff
@@ -1,13 +1,79 @@
...
+      return await collection.doc(id).delete();
+    },
+    async getUsers({ commit }) {
+      try {
+        const res = await functions.httpsCallable('adminGetUsers')();
+        commit('setUsers', res.data);
+      } catch (error) {
+        commit('setUsers', []);
+      }
+    },
```
Comment (warrensearle) [inline src/store/users.js:53]: As discussed please check if this is still needed 😄

**Review by warrensearle — CHANGES_REQUESTED** [review]
All looks great, just one thing to check (see comment)

**Review by NormaJAC — COMMENTED** [review]
ok

**Review by warrensearle — APPROVED** [review]
Code looks great 👍 
Have I mentioned before how good it is to get rid of the cloud function that was being called upon login (the one that meant the UI took ~10 seconds to load) 😃

**Comment by nickaddy** [comment]
@tomlovesgithub @warrensearle I cannot see Users or Roles on Admin, is this a permissions matter? Help!

**Comment by warrensearle** [comment]
> @tomlovesgithub @warrensearle I cannot see Users or Roles on Admin, is this a permissions matter? Help!

Yes, sounds like a permissions matter. Will check and DM you

**Comment by nickaddy** [comment]
@HalcyonJAC I cannot see Users/roles on Develop or Production Admin whether logged in with .gov.uk or .digital address, could you have a look at my permissions pls?

**Comment by warrensearle** [comment]
Hey @HalcyonJAC @nickaddy I just checked our code and Nick would need to be added to the following file:

`src/helpers/authUsers.js`

However could we take the opportunity here to ensure that this is removed from our code base. The Users link showing, or not, should be dependent upon the current user having the correct user role/permission.
We need to get rid of the hard-coded email addresses 🤓

**Comment by HalcyonJAC** [comment]
> Hey @HalcyonJAC @nickaddy I just checked our code and Nick would need to be added to the following file:
> 
> `src/helpers/authUsers.js`
> 
> However could we take the opportunity here to ensure that this is removed from our code base. The Users link showing, or not, should be dependent upon the current user having the correct user role/permission. We need to get rid of the hard-coded email addresses 🤓

@warrensearle The hard-coded email addresses have been removed. Could you re-review it, please?

**Review by warrensearle — APPROVED** [review]
Looks good to me. Just the `UserInvitations` change to include

**Review by tomlovesgithub — APPROVED** [review]

**Review by nickaddy — CHANGES_REQUESTED** [review]
Great to see this feature taking shape! I can see the Users tab now 😊

Users
I can’t confirm all users are listed but I did some random searches and Sophie Austin does not appear to be included – this will need a full audit
Search functionality generally works but when I enter ‘addy’ there are no results returned, when I enter ‘nick’ I get both my entries and one for another Nick, in other words, as a search term, Nick works but not Addy
Pagination works, but the first time I went in, there was a previous button on the first page which took me to an error page
When I click Disable, it changes to Enable and vice versa
I tried to delete Abdalla bidu digital address, received an are you sure pop-up and successfully deleted him (at least the entry disappeared from the list)
I was able to change the role of a user

Roles
I can see all roles
I was able to change permissions of several roles
Able to create a new role, but once I click on Save and Set permissions, screen defaults to the role I was on when I pressed the ‘create a new role’ button rather than the new role
I was able to set permissions on the new role

User Invitations
I have created an invitation (Zubair Ali) but not sure how to test this using MS login?


Question:
Is there a reason why .gov.uk accounts names appear surname, firstname and .digital appear firstname surname?

**Comment by HalcyonJAC** [comment]
Roles
- The screen will now load the new role if you just create a new one.

User Invitations
- I have created an invitation (Zubair Ali) but not sure how to test this using MS login?
    Ans: Would it be possible to get a Microsoft account? 

Question:
- Is there a reason why .gov.uk accounts names appear surname, firstname and .digital appear firstname surname?
    Ans: I just populate the display name from their Microsoft/Google account when they log in. I am not sure if I could change their name.

**Comment by nickaddy** [comment]
@HalcyonJAC Thanks for the updates. There are some other associated tasks before we can go live with this - I'll pick them up once I'm back from leave.

**Review by drieJAC — APPROVED** [review]

**Review by nickaddy — APPROVED** [review]

**Pull request merged**